File: mobile/akd_app/lib/models/quest.dart
/// Quest master record and DailyQuest per-user instance.
class Quest {
  final String id;
  final String title;
  final String description;
  final String? category;
  final String difficultyHint;
  final int rewardPoints;
  final bool active;
  final String createdAt;

  Quest({
    required this.id,
    required this.title,
    required this.description,
    this.category,
    this.difficultyHint = 'easy',
    this.rewardPoints = 10,
    this.active = true,
    required this.createdAt,
  });

  factory Quest.fromJson(Map<String, dynamic> json) {
    return Quest(
      id: json['id'],
      title: json['title'] ?? '',
      description: json['description'] ?? '',
      category: json['category'],
      difficultyHint: json['difficulty_hint'] ?? 'easy',
      rewardPoints: json['reward_points'] ?? 10,
      active: json['active'] ?? true,
      createdAt: json['created_at'] ?? '',
    );
  }
}

class DailyQuest {
  final String id;
  final String userId;
  final String questId;
  final Quest? quest;
  final String date;
  final String status; // assigned | pending_review | completed | rejected
  final String? proofId;
  final bool rewardAwarded;

  DailyQuest({
    required this.id,
    required this.userId,
    required this.questId,
    this.quest,
    required this.date,
    this.status = 'assigned',
    this.proofId,
    this.rewardAwarded = false,
  });

  factory DailyQuest.fromJson(Map<String, dynamic> json) {
    return DailyQuest(
      id: json['id'],
      userId: json['user_id'] ?? '',
      questId: json['quest_id'] ?? '',
      quest: json['quest'] != null ? Quest.fromJson(json['quest']) : null,
      date: json['date'] ?? '',
      status: json['status'] ?? 'assigned',
      proofId: json['proof_id'],
      rewardAwarded: json['reward_awarded'] ?? false,
    );
  }

  bool get isCompleted => status == 'completed';
  bool get isPendingReview => status == 'pending_review';
  bool get isAssigned => status == 'assigned';

  String get statusLabel {
    switch (status) {
      case 'completed':
        return 'Completed ✓';
      case 'pending_review':
        return 'Under Review';
      case 'rejected':
        return 'Rejected';
      default:
        return 'To Do';
    }
  }
}

File: backend/app/models/quest.py
"""Quest and DailyQuest models."""
import uuid
from datetime import datetime, timezone

from app.extensions import db


class Quest(db.Model):
    """Master quest catalog — created/managed by admins."""
    __tablename__ = 'quests'

    id = db.Column(db.String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    title = db.Column(db.String(200), nullable=False)
    description = db.Column(db.Text, nullable=False)
    category = db.Column(db.String(50), nullable=True)  # e.g. "community", "self-care", "environment"
    difficulty_hint = db.Column(db.String(20), default='easy')  # easy / medium / hard
    reward_points = db.Column(db.Integer, default=10, nullable=False)
    reward_gems = db.Column(db.Integer, default=5, nullable=False)
    active = db.Column(db.Boolean, default=True, nullable=False)
    created_by_admin = db.Column(db.String(36), db.ForeignKey('users.id'), nullable=True)
    created_at = db.Column(db.DateTime, default=lambda: datetime.now(timezone.utc), nullable=False)

    daily_instances = db.relationship('DailyQuest', backref='quest', lazy='dynamic')

    def to_dict(self):
        return {
            'id': self.id,
            'title': self.title,
            'description': self.description,
            'category': self.category,
            'difficulty_hint': self.difficulty_hint,
            'reward_points': self.reward_points,
            'reward_gems': self.reward_gems,
            'active': self.active,
            'created_at': self.created_at.isoformat(),
        }


class DailyQuest(db.Model):
    """Per-user daily quest instance."""
    __tablename__ = 'daily_quests'

    id = db.Column(db.String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    user_id = db.Column(db.String(36), db.ForeignKey('users.id'), nullable=False, index=True)
    quest_id = db.Column(db.String(36), db.ForeignKey('quests.id'), nullable=False)
    date = db.Column(db.Date, nullable=False, index=True)
    status = db.Column(db.String(20), default='assigned', nullable=False)
    # status: assigned | pending_review | completed | rejected
    proof_id = db.Column(db.String(36), db.ForeignKey('proofs.id'), nullable=True)
    reward_awarded = db.Column(db.Boolean, default=False, nullable=False)

    __table_args__ = (
        db.UniqueConstraint('user_id', 'quest_id', 'date', name='uq_user_quest_date'),
    )

    def to_dict(self):
        return {
            'id': self.id,
            'user_id': self.user_id,
            'quest_id': self.quest_id,
            'quest': self.quest.to_dict() if self.quest else None,
            'date': self.date.isoformat(),
            'status': self.status,
            'proof_id': self.proof_id,
            'reward_awarded': self.reward_awarded,
        }
